spec/unit/application/device_spec.rb

```diff
@@ -1,564 +1,594 @@
-      allow(Puppet::Util::NetworkDevice).to receive(:init)
-      expect(File).to receive(:file?).and_return(true)
-
-      allow(::File).to receive(:directory?).and_return(false)
-      state_path = tmpfile('state')
-      Puppet[:statedir] = state_path
-      expect(File).to receive(:directory?).with(state_path).and_return(true)
-      expect(FileUtils).not_to receive(:mkdir_p).with(state_path)
-
-      expect(Puppet::Util::CommandLine).to receive(:new).once
-      expect(Puppet::Application::Apply).to receive(:new).once
-
-      expect(Puppet::Configurer).not_to receive(:new)
-      expect { @device.main }.to exit_with 1
+    it "exits if the device list is empty" do
+      expect { device.main }.to exit_with 1
     end
 
-    it "should run an apply, and create the state folder" do
-      allow(@device.options).to receive(:[]).with(:apply).and_return('file.pp')
-      allow(@device.options).to receive(:[]).with(:target).and_return('device1')
-      device_hash = {
-        "device1" => OpenStruct.new(:name => "device1", :url => "ssh://user:pass@testhost", :provider => "cisco"),
+    context 'with some devices configured' do
+      let(:configurer) { instance_double(Puppet::Configurer, 'configurer') }
+      let(:device_hash) {
+        {
+          "device1" => OpenStruct.new(:name => "device1", :url => "ssh://user:pass@testhost", :provider => "cisco"),
+          "device2" => OpenStruct.new(:name => "device2", :url => "https://user:pass@testhost/some/path", :provider => "rest"),
+        }
       }
-      expect(Puppet::Util::NetworkDevice::Config).to receive(:devices).and_return(device_hash)
-      allow(Puppet::Util::NetworkDevice).to receive(:init)
-      expect(File).to receive(:file?).and_return(true)
-      expect(FileUtils).to receive(:mkdir_p).once
 
-      expect(Puppet::Util::CommandLine).to receive(:new).once
-      expect(Puppet::Application::Apply).to receive(:new).once
-
-      expect(Puppet::Configurer).not_to receive(:new)
-      expect { @device.main }.to exit_with 1
-    end
-
-    it "should exit if the device list is empty" do
-      expect { @device.main }.to exit_with 1
-    end
-
-    describe "for each device" do
-      before(:each) do
+      before do
         Puppet[:vardir] = make_absolute("/dummy")
         Puppet[:confdir] = make_absolute("/dummy")
         Puppet[:certname] = "certname"
-        @device_hash = {
-          "device1" => OpenStruct.new(:name => "device1", :url => "ssh://user:pass@testhost", :provider => "cisco"),
-          "device2" => OpenStruct.new(:name => "device2", :url => "https://user:pass@testhost/some/path", :provider => "rest"),
-        }
-        allow(Puppet::Util::NetworkDevice::Config).to receive(:devices).and_return(@device_hash)
+
         allow(Puppet).to receive(:[]=)
         allow(Puppet.settings).to receive(:use)
-        allow(@device).to receive(:setup_host)
+
+        allow(device).to receive(:setup_context)
```

Comment:
It's preferrable to not stub (effectively) private methods of the class we're testing. Better to stub out the state machine's `ensure_client_certificate` method

Reply:
but since this is the mergeup, better to do that kind of change in the original branch